src/Castle.Core/DynamicProxy/Generators/BaseProxyGenerator.cs

```diff
@@ -415,9 +415,10 @@ protected Type ObtainProxyType(CacheKey cacheKey, Func<string, INamingScope, Typ
 				var proxyType = factory.Invoke(name, Scope.NamingScope.SafeSubScope());
 
 				// Upgrade the lock to a write lock. 
-				locker.Upgrade();
-
-				AddToCache(cacheKey, proxyType);
+				using (locker.Upgrade())
```

Comment:
For those reading at home, `SlimReadWriteLock.ForReadingUpgradeable` returns a `SlimUpgradeableReadLockHolder`, `SlimUpgradeableReadLockHolder.Upgrade` creates and holds on to a `SlimWriteLockHolder`. If you don't dispose the writer lock holder the upgradeable read lock holder will do it for you (resulting in `ReaderWriterLockSlim.ExitWriteLock` being called), so this change semantically has no change but makes sense because it is more explicit to scope the write lock's lifetime.